### PR TITLE
Add Adversary.batch_c15n

### PR DESCRIPTION
**Pull request by mzweilin**
# What does this PR do?

`Adversary` assumes batches look like `(input, target)`, but target models may work on very different forms of batches.

This PR adds a configurable `Adversary.batch_c15n()` that converts raw batches into the canonical form `(input, target)`, so `Adversary` can extract `input` and `target`. `Adversary.batch_c15n.revert()` would convert batches back into the original form before feeding the batch to the target model.

The `batch_c15n` not only converts list/tuple/dict input into the canonical tuple and vice versa, but also supports transform/untransform on input/target/batch. The flexibility allows us to reuse the same Adversary for attacking very different models in external projects. For example, we can denormalize `input` in attacking Anomalib models.

## Type of change

Please check all relevant options.

- [ ] Improvement (non-breaking)
- [ ] Bug fix (non-breaking)
- [x] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Testing

Please describe the tests that you ran to verify your changes. Consider listing any relevant details of your test configuration.

- [x] `pytest`
- [x] `CUDA_VISIBLE_DEVICES=0 python -m mart experiment=CIFAR10_CNN_Adv trainer=gpu trainer.precision=16` reports 70% (21 sec/epoch).
- [ ] `CUDA_VISIBLE_DEVICES=0,1 python -m mart experiment=CIFAR10_CNN_Adv trainer=ddp trainer.precision=16 trainer.devices=2 model.optimizer.lr=0.2 trainer.max_steps=2925 datamodule.ims_per_batch=256 datamodule.world_size=2` reports 70% (14 sec/epoch).

## Before submitting

- [x] The title is **self-explanatory** and the description **concisely** explains the PR
- [x] My **PR does only one thing**, instead of bundling different changes together
- [ ] I list all the **breaking changes** introduced by this pull request
- [x] I have commented my code
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have run pre-commit hooks with `pre-commit run -a` command without errors

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Review by dxoigmn — CHANGES_REQUESTED** [review]
You'll probably want to rename this `batch_canonicalization` or `batch_c18n`. Also, since you control what you feed as the `train_dataloader` to `Adversary`, why not canonicalize on `batch` being a `dict` instead of a `tuple`? That feels infinitely more flexible. Will you also get rid of `NormalizedAdversaryAdapter`?

**Comment by mzweilin** [comment]
> You'll probably want to rename this `batch_canonicalization` or `batch_c18n`. Also, since you control what you feed as the `train_dataloader` to `Adversary`, why not canonicalize on `batch` being a `dict` instead of a `tuple`? That feels infinitely more flexible. Will you also get rid of `NormalizedAdversaryAdapter`?

Thanks for the suggestion. I have renamed it to `batch_c15n` for simplicity.

I think it's good to keep the tuple `(input, target)` as the canonical batch form, because the two are required by many sub-components in `Adversary`, such as `Enforcer`. Making it a dictionary may create an illusion that the two parameters are not required.

I would like to keep `NormalizedAdversaryAdapter` for comparison with other adversary implementations, but I am going to update it in a separate PR to match the revised interface in `Adversary`.

File: mart/configs/attack/classification_fgsm_linf.yaml [hunk 1]
```diff
@@ -6,6 +6,7 @@ defaults:
   - gradient_modifier: sign
   - gain: cross_entropy
   - objective: misclassification
+  - batch_c15n: list
```
Comment (dxoigmn) [inline mart/configs/attack/classification_fgsm_linf.yaml:9]: why is this not `tuple`?

File: mart/configs/attack/classification_pgd_linf.yaml [hunk 2]
```diff
@@ -6,6 +6,7 @@ defaults:
   - gradient_modifier: sign
   - gain: cross_entropy
   - objective: misclassification
+  - batch_c15n: list
```
Comment (dxoigmn) [inline mart/configs/attack/classification_pgd_linf.yaml:9]: why is this not `tuple`?

File: mart/attack/adversary.py [hunk 3]
```diff
@@ -165,10 +168,12 @@ def configure_gradient_clipping(
 
     @silent()
     def forward(self, *, batch: torch.Tensor | list | dict, model: Callable):
-        # TODO: We may see different batch format when attacking models outside MART. Add a batch_transform?
-        input, target = batch
+        # Extract and canonicalize input/target so that is convenient for Adversary.
+        input, target = self.batch_c15n(batch)
```
Comment (dxoigmn) [inline mart/attack/adversary.py:172]: Why not just change the signature of `forward` to just include the arguments you need?
```
def forward(self, input, target, *, model):
```
?

**Review by dxoigmn — COMMENTED** [review]
Why don't the batch canonicalizers live in the adversarial training instead of the adversary?

File: mart/configs/attack/classification_fgsm_linf.yaml (hunk 1 above)
Comment (mzweilin) [reply mart/configs/attack/classification_fgsm_linf.yaml:9]: It shouldn't really matter, but `torch.utils.data.Dataloader` returns a `list` instead of `tuple` by default: https://github.com/pytorch/pytorch/blob/9d2ffc5dfa783cf13f8cb49e6cef05d3c6335e76/torch/utils/data/_utils/collate.py#L71

However, we wrote a custom `collate_fn()` for coco, which gives `tuple`. That's why the difference.
https://github.com/IntelLabs/MART/blob/df1734fdbcec9c318575d45c3396f44fbba38959/mart/datamodules/coco.py#L74

File: mart/configs/attack/classification_pgd_linf.yaml (hunk 2 above)
Comment (mzweilin) [reply mart/configs/attack/classification_pgd_linf.yaml:9]: ditto

File: mart/attack/adversary.py (hunk 3 above)
Comment (mzweilin) [reply mart/attack/adversary.py:172]: It worked until Armory gives three data components `x`, `y` and `y_patch_metadata`.

The other benefit is that we can just pass through batch in the adversarial training callback without assuming any structure: https://github.com/IntelLabs/MART/blob/df1734fdbcec9c318575d45c3396f44fbba38959/mart/callbacks/adversary_connector.py#L70

Comment (dxoigmn) [reply mart/attack/adversary.py:172]: How does Armory call `Adversary`? And why can't `AdversaryConnector` just call `batch_c15n` to get the batch into the correct format for `Adversary`?

**Review by dxoigmn — COMMENTED** [review]
`Adversary`'s `forward` is basically:

```python
def forward(batch):
args = unpack_batch(batch)
# Modify args to make them adversarial
adv_batch = pack_batch(args)
```

I still don't get why `Adversary` should be doing this work. Why not move batch unpacking/packing logic into `AdversaryConnector` and `AdversaryWrapper`?

File: mart/attack/adversary.py (hunk 3 above)
Comment (mzweilin) [reply mart/attack/adversary.py:172]: We can definitely canonicalize the batch in the Armory wrapper or `AdversaryConnector`, if we also modify the model to understand the canonical input format (basically inserting `batch_c15n.revert()` before the model).

Armory needs a wrapper to run MART attacks, because we have to either extract or convert their model to a pure-PyTorch model. We shouldn't need a wrapper if the target model is implemented in PyTorch.

The benefit of integrating `batch_c15n` into `Adversary` is that we have a fixed path for the batch canonicalization stuff no matter what models we are targeting at. But I am fine with decoupling it from `Adversary` if you insist.

File: mart/callbacks/adversary_connector.py [hunk 4]
```diff
@@ -67,6 +67,7 @@ def on_after_batch_transfer(self, pl_module, batch, dataloader_idx):
         adversary.to(pl_module.device)
 
         # Directly pass batch instead of assuming it has a structure.
-        batch_adv = adversary(batch=batch, model=pl_module)
+        adversary.fit(batch=batch, model=pl_module)
+        batch_adv = adversary.forward(batch=batch)
```
Comment (dxoigmn) [inline mart/callbacks/adversary_connector.py:71]: Just call adversary?

File: mart/attack/adversary.py [hunk 5]
```diff
@@ -164,11 +164,13 @@ def configure_gradient_clipping(
...
-        input, target = batch
+    def fit(self, *, batch: torch.Tensor | list | dict, model: Callable):
+        # Extract and canonicalize input/target so that is convenient for Adversary.
+        input, target = self.batch_c15n(batch)
 
-        # The attack needs access to the model at every iteration.
+        # Canonical form of batch in the adversary's optimization loop.
+        # We only see the canonicalized input/target in the attack optimization loop.
+        # The attack also needs access to the model at every iteration.
         batch_and_model = (input, target, model)
```
Comment (dxoigmn) [inline mart/attack/adversary.py:174]: I think you can change this to `batch_and_model = (batch, model)` instead now?

File: mart/attack/adversary.py [hunk 6]
```diff
@@ -114,13 +118,9 @@ def configure_optimizers(self):
     def training_step(self, batch_and_model, batch_idx):
         input, target, model = batch_and_model
```
Comment (dxoigmn) [inline mart/attack/adversary.py:119]: This can be `batch, model = batch_and_model` if you make the change below.

File: mart/attack/adversary.py [hunk 7]
```diff
@@ -114,13 +118,9 @@ def configure_optimizers(self):
...
 
-        # Compose input_adv from input, then give to model for updated gain.
-        perturbation = self.perturber(input=input, target=target)
-        input_adv = self.composer(perturbation, input=input, target=target)
-
         # Target model expects input in the original format.
-        # TODO: We assume batch is a tuple, but it may be different for models outside MART. Add a reverse transform?
-        batch_adv = (input_adv, target)
+        # We just compose adversarial examples but not enforce the threat model until after fit().
+        batch_adv = self.forward(input=input, target=target, enforce=False)
```
Comment (dxoigmn) [inline mart/attack/adversary.py:123]: This can become `self.forward(batch, enforce=False)` if you make the changes described.

File: mart/attack/adversary.py [hunk 8]
```diff
@@ -179,15 +181,20 @@ def forward(self, *, batch: torch.Tensor | list | dict, model: Callable):
         self.attacker.fit_loop.max_epochs += 1
         self.attacker.fit(self, train_dataloaders=cycle([batch_and_model]))
 
-        # Get the input_adv for enforcer checking.
+    def forward(self, *, batch=None, input=None, target=None, enforce=True):
```
Comment (dxoigmn) [inline mart/attack/adversary.py:184]: You should be able to change this signature to `def forward(self, batch, enforce=True)`

File: mart/attack/adversary.py [hunk 9]
```diff
@@ -179,15 +181,20 @@ def forward(self, *, batch: torch.Tensor | list | dict, model: Callable):
...
+        """Compose adversarial examples and revert to the original input format."""
+        if batch is not None:
```
Comment (dxoigmn) [inline mart/attack/adversary.py:186]: This controls flow goes away too since we always pass a batch.

File: mart/attack/adversary.py [hunk 10]
```diff
@@ -179,15 +181,20 @@ def forward(self, *, batch: torch.Tensor | list | dict, model: Callable):
...
+
+        # Get the canonicalized input_adv for enforcer checking.
         perturbation = self.perturber(input=input, target=target)
         input_adv = self.composer(perturbation, input=input, target=target)
-        self.enforcer(input_adv, input=input, target=target)
 
-        # TODO: We assume batch is a tuple, but it may be different for models outside MART. Add a reverse transform?
-        # Revert to the original format of batch.
-        batch_adv = (input_adv, target)
+        if enforce:
```
Comment (dxoigmn) [inline mart/attack/adversary.py:193]: Why not always enforce?

**Review by dxoigmn — COMMENTED** [review]
Reviewed on my phone. I think the suggested changes might make it more apparent where c18n should happen given the new Adversary.fit API.

File: mart/attack/adversary.py (hunk 10 above)
Comment (mzweilin) [reply mart/attack/adversary.py]: Fixed in [fd0e847](https://github.com/IntelLabs/MART/pull/215/commits/fd0e8472c08e9e9930c8240adef4f5b1e92e5812)

Simplicity over (negligible) efficiency.

File: mart/attack/adversary.py (hunk 8 above)
Comment (mzweilin) [reply mart/attack/adversary.py]: Fixed in [52d3b23](https://github.com/IntelLabs/MART/pull/215/commits/52d3b23174b5e34fb5b18639fb845477e939bae3)

File: mart/attack/adversary.py (hunk 6 above)
Comment (mzweilin) [reply mart/attack/adversary.py]: [52d3b23](https://github.com/IntelLabs/MART/pull/215/commits/52d3b23174b5e34fb5b18639fb845477e939bae3)

File: mart/attack/adversary.py (hunk 5 above)
Comment (mzweilin) [reply mart/attack/adversary.py]: Fixed in [52d3b23](https://github.com/IntelLabs/MART/pull/215/commits/52d3b23174b5e34fb5b18639fb845477e939bae3)

File: mart/callbacks/adversary_connector.py (hunk 4 above)
Comment (mzweilin) [reply mart/callbacks/adversary_connector.py]: Fixed in 102ef988704fc97de0fb1dac38dcfd9ec5eb1e4c

File: mart/attack/adversary.py (hunk 7 above)
Comment (mzweilin) [reply mart/attack/adversary.py]: Fixed.

File: mart/attack/adversary.py [hunk 11]
```diff
@@ -164,12 +163,12 @@ def configure_gradient_clipping(
...
     @silent()
-    def forward(self, *, batch: torch.Tensor | list | dict, model: Callable):
-        # TODO: We may see different batch format when attacking models outside MART. Add a batch_transform?
-        input, target = batch
-
-        # The attack needs access to the model at every iteration.
-        batch_and_model = (input, target, model)
+    def fit(self, *, batch: torch.Tensor | list | dict, model: Callable):
+        # Extract and canonicalize input for initializing perturbation.
+        # TODO: Get rid of batch_c15n() here by converting perturbation to UninitializedParameter.
```
Comment (mzweilin) [inline mart/attack/adversary.py:168]: This may not be trivial for non-tensor input, such as list and dictionary.

Comment (dxoigmn) [reply mart/attack/adversary.py:168]: Yeah you can't get rid of it here I think.

Comment (dxoigmn) [reply mart/attack/adversary.py:168]: Well maybe you can use `on_train_batch_start` when `batch_idx == 0` to call `self.configure_perturbation()`?

**Review by dxoigmn — APPROVED** [review]
LGTM!
